src/OpenDebugAD7/Tracepoint.cs

```diff
@@ -0,0 +1,509 @@
+        }
+
+        internal static Tracepoint CreateTracepoint(string logMessage)
+        {
+            return new Tracepoint(logMessage); ;
+        }
+
+        internal string GetLogMessage(IDebugThread2 pThread, uint radix, string processName)
+        {
+            string message = LogMessage;
+
+            // There is strings to interpolate in the log message.
+            if (m_indexToExpressions.Count != 0)
+            {
+                message = GetInterpolatedLogMessage(message, pThread, radix, processName);
+            }
+
+            return message;
+        }
+
+        private string GetInterpolatedLogMessage(string logMessage, IDebugThread2 pThread, uint radix, string processName)
+        {
+            if (pThread == null)
+            {
+                return logMessage;
+            }
+
+            // Get topFrame
+            IEnumDebugFrameInfo2 frameInfoEnum;
+            int hr = pThread.EnumFrameInfo(enum_FRAMEINFO_FLAGS.FIF_FRAME | enum_FRAMEINFO_FLAGS.FIF_FLAGS, Constants.EvaluationRadix, out frameInfoEnum);
+            if (hr < 0)
+            {
+                return logMessage;
+            }
+
+            FRAMEINFO[] topFrame = new FRAMEINFO[1];
+            uint fetched = 0;
+            hr = frameInfoEnum.Next(1, topFrame, ref fetched);
+            if (hr < 0 || fetched != 1 || topFrame[0].m_pFrame == null)
+            {
+                return logMessage;
+            }
+
+            Dictionary<string, string> seenExpressions = new Dictionary<string, string>();
+            StringBuilder sb = new StringBuilder();
+            int currIndex = 0;
+
+            foreach (KeyValuePair<int, string> keyValuePair in m_indexToExpressions)
+            {
+                sb.Append(LogMessage.Substring(currIndex, keyValuePair.Key - currIndex));
+
+                currIndex = keyValuePair.Key + keyValuePair.Value.Length;
+
+                if (!seenExpressions.TryGetValue(keyValuePair.Value, out string value))
+                {
+                    if (keyValuePair.Value[0] == '$')
+                    {
+                        value = InterpolateToken(keyValuePair.Value, pThread, topFrame[0].m_pFrame, radix, processName);
+                    }
+                    else
+                    {
+                        string toInterpolate = keyValuePair.Value;
+                        value = InterpolateVariable(toInterpolate.Substring(1, toInterpolate.Length - 2), topFrame[0].m_pFrame, radix);
+                    }
+
+                    // Cache expression
+                    seenExpressions[keyValuePair.Value] = value;
+                }
+                sb.Append(value);
+            }
+
+            sb.Append(LogMessage.Substring(currIndex, LogMessage.Length - currIndex));
+
+            return sb.ToString();
+        }
+
+        private string InterpolateToken(string token, IDebugThread2 pThread, IDebugStackFrame2 topFrame, uint radix, string processName)
+        {
+            switch (token)
```

Comment:
can I use lowercase tokens?

Reply:
No, these tokens are the same as the one in VS which are all uppercase.
![image](https://user-images.githubusercontent.com/3953714/85642520-a034c400-b646-11ea-9727-9e9ecd7667d5.png)

https://docs.microsoft.com/en-us/visualstudio/debugger/using-tracepoints?view=vs-2019#actions-menu